lib/libzutil/zutil_import.c

```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
 
 	if (lio_listio(LIO_WAIT, aiocbps, VDEV_LABELS, NULL) != 0) {
 		int saved_errno = errno;
-
-		if (errno == EAGAIN || errno == EINTR || errno == EIO) {
-			/*
-			 * A portion of the requests may have been submitted.
-			 * Clean them up.
-			 */
-			for (l = 0; l < VDEV_LABELS; l++) {
-				errno = 0;
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
+		errno = 0;
+		int do_slow = 0;
+		error = -1;
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			switch (aio_error(&aiocbs[l])) {
+				case 0:
+					break;
+				case -1:
+					break;
+				case EOPNOTSUPP:
+					do_slow = 1;
+				case EIO:
+				case EINTR:
+				case EAGAIN:
+					/*
+					 * A portion of the requests may have
+					 * been submitted. Clean them up.
+					 */
 					(void) aio_return(&aiocbs[l]);
 			}
 		}
+		if (do_slow) {
+			/*
+			 * At least some IO involved access unsafe-for-AIO
+			 * files. Let's try again, without AIO this time.
+			 */
+			error = zpool_read_label_slow(fd, config, num_labels);
+			saved_errno = errno;
+		}
 		free(labels);
 		errno = saved_errno;
-		return (-1);
+		return (error);
```

Comment:
Why are you changing the return value?  That doesn't seem related to the rest of the PR.

Reply:
The goal was to not return -1 if zpool_read_label_slow gets called and returns 0, since we did succeed at the purpose of the function.

Reply:
Ok, and `zpool_read_label_slow` will only return either 0, or -1.  I get it.  I thought that you had changed it to return an errno or something, but now I see that it is not the case.